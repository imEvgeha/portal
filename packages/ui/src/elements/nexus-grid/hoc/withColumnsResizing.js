import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {setGridColumnsSize} from '../nexusGridActions';
import {createColumnsSizeSelector} from '../nexusGridSelectors';

const AG_GRID_DEF_COL_DEF = {
    resizable: true,
};

const withColumnsResizing = ({colDef = AG_GRID_DEF_COL_DEF, defaultColDef} = {}) => WrappedComponent => {
    const ComposedComponent = props => {
        const {columnDefs, id, existingColumnsSize, updateGridColumnsSize, defaultColDef} = props;
        const existingTableColumnsSize = existingColumnsSize[id] || {};
        let columnDefsWithSizes = columnDefs;
        if (existingTableColumnsSize) {
            columnDefsWithSizes = columnDefs.map(c =>
                c.colId && existingTableColumnsSize[c.colId] ? {...c, width: existingTableColumnsSize[c.colId]} : c
            );
        }
        const [columnsSize, setColumnsSize] = useState(existingTableColumnsSize);

        const handleColumnResized = e => {
            if (e.finished && e.column) {
                const updatedColumnsSize = {...columnsSize, [e.column.colDef.colId]: e.column.actualWidth};
                setColumnsSize(updatedColumnsSize);
                id && updateGridColumnsSize(id, updatedColumnsSize);
            }
        };

        return (
            <WrappedComponent
                {...props}
                defaultColDef={{
                    ...colDef,
                    ...defaultColDef,
                }}
                onColumnResized={handleColumnResized}
                columnDefs={columnDefsWithSizes}
            />
        );
    };

    const createMapStateToProps = () => {
        const columnsSizeSelector = createColumnsSizeSelector();
        return (state, props) => ({
            existingColumnsSize: columnsSizeSelector(state, props),
        });
    };

    const mapDispatchToProps = dispatch => ({
        updateGridColumnsSize: (id, payload) => dispatch(setGridColumnsSize(id, payload)),
    });

    ComposedComponent.propTypes = {
        ...WrappedComponent.propTypes,
        updateGridColumnsSize: PropTypes.func,
        existingColumnsSize: PropTypes.object,
    };

    ComposedComponent.defaultProps = {
        ...WrappedComponent.defaultProps,
        updateGridColumnsSize: null,
        existingColumnsSize: {},
    };

    // eslint-disable-next-line
    return connect(createMapStateToProps, mapDispatchToProps)(ComposedComponent);
};

export default withColumnsResizing;
