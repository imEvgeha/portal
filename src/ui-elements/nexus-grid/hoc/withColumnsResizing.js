import React, {useEffect, useState} from 'react';
import {store} from '../../../index';
import {setGridColumnsSize} from '../../../stores/actions';
import {createColumnsSizeSelector} from '../../../stores/selectors/columnsSize/columnsSizelSelectors';
import {connect} from 'react-redux';

const AG_GRID_DEF_COL_DEF = {
    resizable: true
};

const withColumnsResizing = (colDef= AG_GRID_DEF_COL_DEF) => WrappedComponent =>{
    const ComposedComponent = (props) => {
        const {columnDefs, id, existingColumnsSize} = props;
        const [columnsSize, setColumnsSize] = useState({});

        useEffect(() => {
            return () => {
                id && store.dispatch(setGridColumnsSize(id, columnsSize));
            };
        }, [columnsSize] );

        const handleColumnResized  = e => {
            if(e.finished){
                setColumnsSize({...columnsSize, [e.column.colDef.colId] : e.column.actualWidth});
            }
        };

        return (
            <WrappedComponent
                {...props}
                defaultColDef={colDef}
                onColumnResized={handleColumnResized}
            />
        );
    };

    const createMapStateToProps = () => {
        const columnsSizeSelector = createColumnsSizeSelector();
        return (state, props) => ({
            existingColumnsSize: columnsSizeSelector(state, props),
        });
    };

    return connect(createMapStateToProps)(ComposedComponent); // eslint-disable-line
};

export default withColumnsResizing;

