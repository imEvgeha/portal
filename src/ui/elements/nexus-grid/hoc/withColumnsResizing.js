import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {setGridColumnsSize} from '../../../../pages/legacy/stores/actions';
import {createColumnsSizeSelector} from '../../../../pages/legacy/stores/selectors/columnsSize/columnsSizelSelectors';

const AG_GRID_DEF_COL_DEF = {
    resizable: true
};

const withColumnsResizing = ({colDef= AG_GRID_DEF_COL_DEF} = {}) => WrappedComponent =>{
     const ComposedComponent = (props) => {
        const {columnDefs, id, existingColumnsSize, updateGridColumnsSize} = props;
        const existingTableColumnsSize = existingColumnsSize[id] || {};
        let columnDefsWithSizes = columnDefs;
        if (existingTableColumnsSize) {
            columnDefsWithSizes = columnDefs.map(c => c.colId && existingTableColumnsSize[c.colId] ? {...c, width: existingTableColumnsSize[c.colId]} : c);
        }
        const [columnsSize, setColumnsSize] = useState(existingTableColumnsSize);

        const handleColumnResized  = e => {
            if (e.finished){
                const updatedColumnsSize = {...columnsSize, [e.column.colDef.colId] : e.column.actualWidth};
                setColumnsSize(updatedColumnsSize);
                id && updateGridColumnsSize(id, updatedColumnsSize);
            }
        };

        return (
            <WrappedComponent
                {...props}
                defaultColDef={colDef}
                onColumnResized={handleColumnResized}
                columnDefs={columnDefsWithSizes}
            />
         );
    };

    const createMapStateToProps = () => {
        const columnsSizeSelector = createColumnsSizeSelector();
        return (state, props) => ({
            existingColumnsSize: columnsSizeSelector(state, props),
        });
    };

    const mapDispatchToProps = dispatch => ({
        updateGridColumnsSize: (id, payload) => dispatch(setGridColumnsSize(id, payload)),
    });

    ComposedComponent.propTypes = {
        ...WrappedComponent.propTypes,
        updateGridColumnsSize: PropTypes.func,
        existingColumnsSize: PropTypes.object,
    };

    ComposedComponent.defaultProps = {
        ...WrappedComponent.defaultProps,
        updateGridColumnsSize: null,
        existingColumnsSize: {},
    };

    return connect(createMapStateToProps, mapDispatchToProps)(ComposedComponent); // eslint-disable-line
};

export default withColumnsResizing;

