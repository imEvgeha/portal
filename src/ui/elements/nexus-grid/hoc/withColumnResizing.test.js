import React from 'react';
import {shallow} from 'enzyme';
import configureStore from 'redux-mock-store';
import withColumnsResizing from './withColumnsResizing';

describe('withColumnsResizing', () => {
    let wrapper;
    let WithLoadingComponent;

    beforeEach(() => {
        const Component = <h1>Test</h1>;
        const props = {
            id: 'rightsRepo',
            columnDefs: [
                {
                    field: 'title',
                    headerName: 'Title',
                    colId: 'title',
                    width: 150
                },
                {
                    field: 'releaseYear',
                    headerName: 'Release Year',
                    colId: 'releaseYear',
                    valueFormatter: null,
                    width: 150
                },
                {
                    field: 'licensor',
                    headerName: 'Licensor',
                    colId: 'licensor',
                    width: 150
                }
            ],
            updateGridColumnsSize: () => null
        };
        WithLoadingComponent =  withColumnsResizing()(Component);


        const mockStore = configureStore();
        let store = mockStore({
            root: {
                columnsSize: {
                    rightsRepo: {
                        title: 368,
                        licensor: 225
                    }
                }
            },
        });
        wrapper = shallow(<WithLoadingComponent store={store} {...props} />);
    });

    it('should render the component', () => {
        expect(wrapper).not.toBe(null);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

});

