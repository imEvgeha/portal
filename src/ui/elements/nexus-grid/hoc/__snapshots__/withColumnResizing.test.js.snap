// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withColumnsResizing should match snapshot 1`] = `
<ComposedComponent
  columnDefs={
    Array [
      Object {
        "colId": "title",
        "field": "title",
        "headerName": "Title",
        "width": 150,
      },
      Object {
        "colId": "releaseYear",
        "field": "releaseYear",
        "headerName": "Release Year",
        "valueFormatter": null,
        "width": 150,
      },
      Object {
        "colId": "licensor",
        "field": "licensor",
        "headerName": "Licensor",
        "width": 150,
      },
    ]
  }
  existingColumnsSize={
    Object {
      "rightsRepo": Object {
        "licensor": 225,
        "title": 368,
      },
    }
  }
  id="rightsRepo"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  updateGridColumnsSize={[Function]}
/>
`;
